src/test/java/com/meta/chatbridge/message/FBMessageHandlerTest.java

```diff
@@ -0,0 +1,368 @@
+              .put("hub.challenge", Integer.toString(challenge))
+              .put("hub.verify_token", token)
+              .build();
+      response = getRequest("testfbmessage", runner.port(), params);
+    }
+    assertThat(response.getCode()).isEqualTo(200);
+    InputStream inputStream = ((BasicClassicHttpResponse) response).getEntity().getContent();
+    byte[] input = new byte[inputStream.available()];
+    inputStream.read(input);
+    String text = new String(input, StandardCharsets.UTF_8);
+    assertThat(text).isEqualTo(Integer.toString(challenge));
+  }
+
+  private static Request createMessageRequest(
+      String body, PipelinesRunner runner, boolean calculateHmac)
+      throws IOException, URISyntaxException {
+    @SuppressWarnings("unchecked") // for the scope of this test this is guaranteed
+    Pipeline<FBMessage> pipeline =
+        (Pipeline<FBMessage>) runner.pipelines().stream().findAny().get();
+    String path = pipeline.path();
+
+    // for the scope of this test this is guaranteed
+    FBMessageHandler messageHandler = (FBMessageHandler) pipeline.messageHandler();
+
+    URI uri =
+        URIBuilder.localhost().setScheme("http").setPort(runner.port()).appendPath(path).build();
+
+    Request request = Request.post(uri).bodyString(body, ContentType.APPLICATION_JSON);
+    if (calculateHmac) {
+      String hmac = messageHandler.hmac(body);
+      request.setHeader("X-Hub-Signature-256", "sha256=" + hmac);
+    }
+
+    return request;
+  }
+
+  private static Request createMessageRequest(String body, PipelinesRunner runner)
+      throws IOException, URISyntaxException {
+    return createMessageRequest(body, runner, true);
+  }
+
+  private Function<Identifier, URI> testURLFactoryFactory(Identifier pageId) {
+    return p -> {
+      assertThat(p).isEqualTo(pageId);
+      try {
+        return URIBuilder.localhost().setScheme("http").setPort(app.port()).build();
+      } catch (URISyntaxException | UnknownHostException e) {
+        fail("failed building url");
+        throw new RuntimeException(e);
+      }
+    };
+  }
+
+  private record TestArgument(
+      String name,
+      int expectedReturnCode,
+      ThrowableFunction<PipelinesRunner, Request> requestFactory,
+      boolean messageExpected) {
+
+    // hacky way to try the send twice scenario
+    int timesToSendMessage() {
+      return name.contains("send it twice") ? 2 : 1;
+    }
+  }
+
+  static Stream<Arguments> requestFactory() throws JsonProcessingException {
+    JsonNode duplicateMessage = MAPPER.readTree(SAMPLE_MESSAGE);
+    ArrayNode messagingArray = (ArrayNode) duplicateMessage.get("entry").get(0).get("messaging");
+    messagingArray.add(messagingArray.get(0));
+
+    JsonNode duplicateEntry = MAPPER.readTree(SAMPLE_MESSAGE);
+    ArrayNode entryArray = (ArrayNode) duplicateMessage.get("entry");
+    entryArray.add(entryArray.get(0));
+    Stream<TestArgument> arguments =
+        Stream.of(
+            new TestArgument(
```

Comment:
Maybe a valid sample with more than 1 entry?

Reply:
it's actually called out in the docs that there will never be more than one entry